packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts

```diff
@@ -591,6 +592,151 @@ async function updateAutoLink(
+    if (line.startsWith('</')) {
+      indent--;
+    }
+    output += '  '.repeat(indent) + line.trim() + '>\r\n';
+    if (line.endsWith('?')) {
+      // header, don't change indent
+    } else if (line.endsWith('/')) {
+      // self-closing tag: <foo />
+    } else if (!line.startsWith('</')) {
+      indent++;
+    }
+  }
+  if (indent !== 0) throw new Error('malformed xml');
+  return output;
+}
+
+function ensureWinUIDialect(
+  slnFile: string,
+  windowsProjectConfig: WindowsProjectConfig,
+  checkMode: boolean,
+) {
+  console.log('ensure dialect');
+  const buildFlagsProps = path.join(path.dirname(slnFile), 'BuildFlags.props');
+  let changesNeeded = false;
+  if (fs.existsSync(buildFlagsProps)) {
+    const buildFlagsContents = configUtils.readProjectFile(buildFlagsProps);
+    const useWinUI3FromConfig = windowsProjectConfig.useWinUI3;
+    const useWinUI3FromBuildFlags =
+      configUtils
+        .tryFindPropertyValue(buildFlagsContents, 'UseWinUI3')
+        ?.toLowerCase() === 'true';
+    // use the UseWinUI3 value in react-native.config.js, or if not present, the value from BuildFlags.props
+    changesNeeded = fixPackagesConfig(
+      useWinUI3FromConfig !== undefined
+        ? useWinUI3FromConfig
+        : useWinUI3FromBuildFlags,
+      windowsProjectConfig,
+      checkMode,
+    );
+    if (useWinUI3FromConfig !== undefined) {
+      // Make sure BuildFlags matches the value that comes from react-native.config.js
+      const node = buildFlagsContents.getElementsByTagName('UseWinUI3');
+      const newValue = useWinUI3FromConfig ? 'true' : 'false';
+      changesNeeded = changesNeeded || node.item(0)?.textContent !== newValue;
+      if (!checkMode && changesNeeded) {
+        node.item(0)!.textContent = newValue;
+        const buildFlagsOutput = new XMLSerializer().serializeToString(
+          buildFlagsContents,
+        );
+        fs.writeFileSync(buildFlagsProps, buildFlagsOutput);
+      }
+    }
+  }
+  return changesNeeded;
+}
+
+function fixPackagesConfig(
+  useWinUI3: boolean,
+  project: WindowsProjectConfig,
+  checkMode: boolean,
+) {
+  const projectFile = path.join(
+    project.folder,
+    project.sourceDir,
+    project.project!.projectFile,
+  );
+  const packagesConfig = path.join(
+    path.dirname(projectFile),
+    'packages.config',
+  );
+  console.log(`UseWinUI = ${useWinUI3}`);
+  console.log(`packages.config = ${packagesConfig}`);
+  let changed = false;
+
+  if (fs.existsSync(packagesConfig)) {
+    // if we don't have a packages.config, then this is a C# project, in which case we use <PackageReference> and dynamically pick the right XAML package.
+    const packagesConfigContents = configUtils.readProjectFile(packagesConfig);
+    const packageElements = packagesConfigContents.documentElement.getElementsByTagName(
+      'package',
+    );
+
+    const winuiPropsContents = configUtils.readProjectFile(
+      require.resolve('react-native-windows/PropertySheets/WinUI.props'),
```

Comment:
This will break for PNPM or repos with multiple RNW versions, since the CLI does not depend on react-native-windows.

Need to search starting from project root. There should be other examples of this in the CLI, and I think this file might even have a `resolveTemplateDirectory` function or something like that.